contracts/implementations/nodes/Integration.sol

```diff
@@ -0,0 +1,287 @@
+
+import {DEFAULT_ADMIN_ROLE} from "../../shared/Roles.sol";
+import {AttributeInfoPair} from "../../shared/Types.sol";
+
+import "@solidstate/contracts/access/access_control/AccessControlInternal.sol";
+
+/// @title Integration
+/// @notice Contract that represents the Integration node
+contract Integration is AccessControlInternal {
+    event IntegrationIdProxySet(address indexed proxy);
+    event IntegrationAttributeAdded(string attribute);
+    event IntegrationAttributeSet(
+        uint256 tokenId,
+        string attribute,
+        string info
+    );
+    event ControllerSet(address indexed controller);
+    event IntegrationNodeMinted(uint256 tokenId, address indexed owner);
+
+    modifier onlyNftProxy() {
+        require(
+            msg.sender == IntegrationStorage.getStorage().idProxyAddress,
+            "Only NFT Proxy"
+        );
+        _;
+    }
+
+    // ***** Admin management ***** //
+
+    /// @notice Sets the NFT proxy associated with the Integration node
+    /// @dev Only an admin can set the address
+    /// @param addr The address of the proxy
+    function setIntegrationIdProxyAddress(address addr)
+        external
+        onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        require(addr != address(0), "Non zero address");
+        IntegrationStorage.getStorage().idProxyAddress = addr;
+
+        emit IntegrationIdProxySet(addr);
+    }
+
+    /// @notice Adds an attribute to the whitelist
+    /// @dev Only an admin can add a new attribute
+    /// @param attribute The attribute to be added
+    function addIntegrationAttribute(string calldata attribute)
+        external
+        onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        require(
+            AttributeSet.add(
+                IntegrationStorage.getStorage().whitelistedAttributes,
+                attribute
+            ),
+            "Attribute already exists"
+        );
+
+        emit IntegrationAttributeAdded(attribute);
+    }
+
+    /// @notice Sets a address controller
+    /// @dev Only an admin can set new controllers
+    /// @param _controller The address of the controller
+    function setIntegrationController(address _controller)
+        external
+        onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        IntegrationStorage.Storage storage s = IntegrationStorage.getStorage();
+        require(_controller != address(0), "Non zero address");
+        require(
+            !s.controllers[_controller].isController,
+            "Already a controller"
+        );
+
+        s.controllers[_controller].isController = true;
+
+        emit ControllerSet(_controller);
+    }
+
+    // ***** Interaction with nodes ***** //
+
+    /// @notice Mints integrations in batch
+    /// @dev Caller must be an admin
+    /// @dev It is assumed the 'Name' attribute is whitelisted in advance
+    /// @param owner The address of the new owner
+    /// @param names List of integration names
+    function mintIntegrationBatch(address owner, string[] calldata names)
```

Comment:
We're minting two of these for now, so this isn't so crucial.

Reply:
Oh but this one isn't strict about controllers. I see.

We should write this pattern down somewhere.